### DISCUSSION TITLE
Stereotype SysML blocks

### DISCUSSION THREAD

**Discussion by tompkins-ct** [gaphor/gaphor#2465] [Q&A] [answered]
How would I stereotype the SysML components (e.g. Proxy Ports or Blocks)?

The metaclass doesn't allow for selecting SysML components and generalization of the block stereotype doesn't create a new stereotype. Creating a stereotype generalized from a new block stereotype results in creating duplicate stereotypes. 

![image](https://github.com/gaphor/gaphor/assets/58485783/b25432c4-f1f2-4757-b6d8-e5cc101bc4f8)

**Comment by mikekidner** [comment]
Also, once a stereotype is created, how is it referenced by the CSS?

**Reply by tompkins-ct** [reply]
@amolenaar @frandieguez  @danyeaw

Any thoughts on this? 

Maybe extending on @mikekidner question. How do we reference stereotypes by CSS and if I were to give a block an attribute say "status" or "draft" can I style all blocks based on an attribute or reference a comment in the note box?

I've noticed I can isolate valuetype with:
```css
datatype[subject*=valuetype] {
  background-color: blue
 }
```
however, it doesn't seem to work with stereotyped blocks (maybe because the class is stereotyped and applied to the block?)

**Comment by amolenaar** [comment, marked as answer]
Something like this? 

![image](https://github.com/gaphor/gaphor/assets/96249/c081324e-0ccb-4953-b44c-920189e82846)

Although Blocks are considered stereotypes of Classes in the SysML spec, we treat them as normal metaclasses -- similar to how one would create a stereotype from a Class.

The CSS I created is:

```css
block[appliedStereotype.classifier.name=hardware] {
  background-color: #99c1f1;
}
```

These are references in the model. The stereotype is referencing an "instance" of the stereotype (since stereotypes are normally part of a meta model).

See the UML spec on [Instances](https://docs.gaphor.org/en/latest/models/uml/03._Classification.html#instances) and [Gaphor Specific Constructs](https://docs.gaphor.org/en/latest/models/uml/A._Gaphor_Specific_Constructs.html).

**Reply by mikekidner** [reply]
That's great! Thank you.

**Reply by tompkins-ct** [reply]
Great. Thanks!

One thing I notice with the stereotypes of SysML components is that they aren't in the dropdown list on the right for the `metaclass`. So, if you use a name for the `metaclass` that isn't in the list, it reverts to `abstraction` which is the first item in the drop-down menu. I don't know if that's the intent for stereotypes of stereotypes or if something else is missing.

![image](https://github.com/gaphor/gaphor/assets/58485783/d61b0d4a-5ed3-49eb-bcf6-9845861e1851)

**Comment by mikekidner** [comment]
Yes I saw that too, but you can rename the element BEFORE connecting the stereotype. And it flows through. 
<img width="307" alt="Screen Shot 2023-07-21 at 8 06 42 am" src="https://github.com/gaphor/gaphor/assets/50702106/3f3d1805-ef29-4dce-bff2-9aaf08d5e01f">
<img width="324" alt="Screen Shot 2023-07-21 at 8 07 35 am" src="https://github.com/gaphor/gaphor/assets/50702106/167a79d4-7b86-45c3-a091-8cdb23cef92e">

**Comment by amolenaar** [comment]
Ouch. I expected all model elements to be present. Or at least have the modeling elements of the active profile available.

It's probably a good idea to support stereotypes for the SysML and RAAML profiles by default.

**Comment by amolenaar** [comment]
I checked the code. Indeed, the list of classes only contains classes from the UML module. Unfortunately the code does not have access to the active modeling profile.

It does make sense to have a _Profile_ section in the SysML toolbox, though.

**Reply by tompkins-ct** [reply]
I think is makes sense to have a Profile section in SysML. I typically make a lot of stereotypes of blocks to organize the overall model and classify components (e.g, standard, system, subsystem, component, assembly, function...)
